### DISCUSSION TITLE
How to use and organize Conda packages

### DISCUSSION THREAD

**Discussion by sapetnioc** [brainvisa/brainvisa.github.io#143] [General]
I wrote this rather long text to discuss the most urgent topics that we must agree on in order to move forward with using Conda to distribute our software. I tried to describe how users and developers could use Conda environment and how we could organize packages. As I often do, I wrote the document as if choices were already done but this is only because it is simpler to write. Everything can be changed. 

# User experience

Mambaforge (i.e. Conda with Mamba installed and without anaconda repository) under Linux is the standard environment for all users. Web accessible documentation describes how to set-up and activate this environment on various platforms with the following options:
    - Linux: install Mambaforge (download and execute sh script). Activate using activate script or conda activate command.
    - Windows: 
        - use VirtulaBox to install Linux and then Mambaforge. We can provide a base image already configured.
        - use WSL2 to install Linux and then Mambaforge.
    - MacOS: use virtualbox to install Linux and then Mambaforge.

Once the Mambaforge environment is activated, user must declare our package repository: `mamba config --add channels https://brainvisa.info/conda`. Packages are then installed with `mamba install` and software are run directly from the shell.

We may provide an contenerized version of the Mambaforge environment for Linux if necessary. I will not talk of this here since I believe it is not urgent. It can only be done once the Conda usage is fully defined.

# Developper experience

Setting up a developpment environment is very similar to a user environment. The only thing that changes are the installed packages. In a developpment environement, user packages must not be installed. Instead, a single package must be installed four our software : `brainvisa-dev-<distro>`. Where `<distro>` is a name corresponding to the default set of brainvisa-cmake components to be configured by default. This corresponds to the `distro` option of casa-distro. Installation of this package does three things:

- install a file `etc/brainvisa/bv_maker.cfg` whose content is related to the selected distro.
- install two shell scripts : `etc/conda/activate.d/brainvisa-dev_activate.sh` and `etc/conda/deactivate.d/brainvisa-dev_deactivate.sh` that sets all necessary environment variables during environment activation. In particular, it sets `PATH` to point to (not yet existing) build tree and sets `BRAINVISA_BVMAKER_CFG=<conda>/etc/brainvisa/bv_maker.cfg`.
- install `bin/brainvisa-dev-setup` script that must be used once to setup the developpement environment.

Once the `brainvisa-dev` package is installed, the developer can modify the `etc/brainvisa/bv_maker.cfg` file if necessary and then run `brainvisa-dev-setup`. This is mainly necessary to download brainvisa-cmake in order to be able to download sources and run the first build. `bv_maker` is now ready to be launched from sources thanks to all the variables defined by `brainvisa-dev_activate.sh` script.

By default sources are be put in the `src` directory and build files are be put in the `build` directory. The build directory is ready to be used in an activated environment because of the variables (especially PATH) set by activation script.

# Name of the base environment

Whatever a user want to do with our software, he will have to start by installing a base enviromnent that is common to all our software. Today this environment is using an Apptainer (i.e. Singularity) container and we are working on switching this environment to Conda. We have to communicate on this environment, especially to document the software installation procedure. Therefore we need a name for this environment. In this document, I chose to name this environment "BrainVISA base" environment and I would like to be sure that we all agree on this. I insist on the importance of the BrainVISA part of the name because other options I considered were Populse and CASA. If we choose BrainVISA as the host name for basic software infrastructure, we should better to be coherent for the naming and hosting of projects included in this base environment (see the next point: Pa).

Here are the reasons why I chose BrainVISA over Populse or CASA:
- We already have a BrainVISA website where we can host all the Conda packages that depends on the base environment. Using BrainVISA is coherent when we ask user to add the Conda channel (i.e. a repository of package) located in https://brainvisa.info/conda.
- BrainVISA is the most known of the three. In my opinion, there are more pros than cons.
- This base environment is mainly developped by the BrainVISA team.

And, in parallel of BrainVISA, I chose "base" because we will have a installable package that all our software will depend on. I wanted to call it `brainvisa-common` but I saw that on conda-forge there are more packages using the `*-base` convention than `*.common`.

# Repositories and packages organization

Conda support repository of packages (called channels in Conda) that are either in a local directory or on a web site. We will use web only for public repositories (i.e. anonymous access without access rights). Private repositories will be on a file system (eventually shared) and will rely on file system rights to manage access.

We will maintain at least a single public repository. We may decide to use the repository system to distribute private content to users. Sharing a repository is an alternative to sharing a whole build tree. We will probably not rush to setup private repositories but to have a complete discussion on package organization, I make the assumption that all package distribution goes through a repository and an installation or update in Conda by end user.

## Repositories

Under the above assumption, we would define the following repositories:

BrainVISA (https://brainvisa.info/conda): public repository containing all base infrastructure packages as well as BrainVISA specific packages. 

CEA (directory rsynced in various facilities): contains packages specific to MirCEN, Neurospin and SHFJ

CATI (directory rsynced in various facilities): contains packages specific to CATI

## Packages

In the following, I list the packages names that are going to be in each repository. For each package, I list the brainvisa-cmake component that are included in the package.

### BrainVISA repository

- brainvisa-base
    - soma-base
    - capsul
    - populse_db
    - soma_workflow

- brainvisa-data-processing: depends on brainvisa-base
    - aims-free
    - aims-gpl
    - brainvisa-share
    - axon

- anatomist: depends on brainvisa-data-processing
    - anatomist-free
    - anatomist-gpl

- brainvisa-disco:
    - brainvisa-disco

- brainvisa-spm:
    - brainvisa-spm

- brainvisa-freesurfer:
    - brainvisa_freesurfer

- brainvisa-cortical-surface:
    - cortical_surface-gpl
    - cortical_surface-nonfree

- brainvisa-highres-cortex:
    - highres-cortex

- morphologist:
    - morpho-deepsulci
    - morphologist
    - morphologist-gpl
    - morphologist-nonfree
    - sulci-nonfree

- brainvisa-dev-{distro}
    - brainvisa-cmake
    - casa-distro

- brainrat:
    - brainrat-gpl, brainrat-private

### CEA repository

- 3dns:
    - 3dns-private

- bioprocessing:
    - bioprocessing

- primatologist:
    - primatologist-gpl

- morphologist-baby:
    - morphologist-baby

- brainvisa-constellation: 
    - constellation-gpl
    - constellation-nonfree

### CATI repository

- cati-neuroimaging:
    - catidb-client
    - deidentification
    - longitudinal_pipelines
    - nuclear_imaging-gpl
    - nuclear_imaging-nonfree
    - rsfmri
    - sacha-gpl
    - sacha-nonfree
    - whasa-gpl
    - whasa-nonfree

**Comment by denisri** [comment]
I'm more or less OK with all that. I have no real feeling for the developer part, and I think it's probably not the first priority (even if external toolboxes developers will get interested in it).
For the packages names and grouping, almost OK for me, except:
- should axon be part of brainvisa-data-processing ? As it is planned to be obsolete and removed one day, we could make it separate (but it's probably not required)
- morphologist-baby is not distributed and should be in the cea repository
- brainrat-gpl and brainrat-private are distributed as a user toolbox in the public repository, but its source code (brainrat-private) is not. Maybe it's not a problem moving them to the public repository. The developer package could perhaps stay in the cea repos.
- constellation is not part of the public repository yet. It could become part of it, but it's not done so far. It is a bit tricky to use, so we wanted to have a good doc with it before shipping it to users.

**Comment by sapetnioc** [comment]
For axon, I hesitated. But I think no one will ever install `brainvisa-axon` but one of its dependent package. Therefore I end up hiding it in `brainvisa-data-processing` so that it can disappear silently on day.

I modified the text to integrate your other points.

**Comment by Hboni** [comment]
That sounds great to me. I was wondering will it be easy to install several "versions" of the environment on one system? As we can do now with casa-distro, and select the environment by launching the good `bv` command.

A similar question is about channels, can we have several versions of one channel installed locally, with ease to switch between versions? 

For the packages, I was thinking about the qualiCATI environment. As qualiCATI is not installed in brainvisa, you did not list it in your packages repartition. But could we imagine having a conda channel with qualiCATI to be installed inside the brainvisa-base environment?

We may also have a discussion about CATI channels organisation, we may want to split cati-neuroimaging. But I think it is not urgent also.
